server/app/repository/ReportingRepositoryFactory.java

```diff
@@ -0,0 +1,54 @@
+package repository;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.ebean.DB;
+import io.ebean.Database;
+import java.time.Clock;
+import javax.inject.Inject;
+import models.ProgramModel;
+import services.program.ProgramDefinition;
+
+/**
+ * Factory for initializing ReportingRepository.
+ *
+ * <p>Avoids n+1 DB queries when acquiring the public name of a program when loading program
+ * reporting statistics.
+ */
+public final class ReportingRepositoryFactory {
+  private final Clock clock;
+  private final Database database;
+  private final VersionRepository versionRepository;
+  private final ProgramRepository programRepository;
+
+  @Inject
+  public ReportingRepositoryFactory(
+      Clock clock, VersionRepository versionRepository, ProgramRepository programRepository) {
+    this.clock = Preconditions.checkNotNull(clock);
+    this.database = DB.getDefault();
+    this.versionRepository = Preconditions.checkNotNull(versionRepository);
+    this.programRepository = Preconditions.checkNotNull(programRepository);
+  }
+
+  /**
+   * Creating a ReportingRepository object with <code>List&lt;ProgramModel&gt;</code> output as a
+   * <code>ImmutableMap</code>.
+   */
+  public ReportingRepository create() {
+    ImmutableList<ProgramModel> listOfPrograms = versionRepository.getActiveVersion().getPrograms();
```

Comment:
@dkatzz  - Is versionRepository needed here? I am trying to get all active program list.

Reply:
Thanks for checking! We should do this by calling `versionRepository.getProgramsForVersion(versionRepository.getActiveVersion()));`. `version.getPrograms()` is an expensive call, so we can use the cache if its available by using the version repository function https://github.com/civiform/civiform/blob/e87bdabb0a49eabdc042edfc314b04f894adfa86/server/app/repository/VersionRepository.java#L528